keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/EnvironmentConstraintRunner.kt

```diff
@@ -229,9 +229,29 @@ class EnvironmentConstraintRunner(
   ): Boolean =
     evaluators.all { evaluator ->
       !environment.hasSupportedConstraint(evaluator) ||
-        evaluator.canPromote(artifact, version, deliveryConfig, environment)
+        evaluator.testConstraint(artifact, version, deliveryConfig, environment)
     }
 
+  private fun ConstraintEvaluator<*>.testConstraint(
+    artifact: DeliveryArtifact,
+    version: String,
+    deliveryConfig: DeliveryConfig,
+    environment: Environment
+  ) =
+    canPromote(artifact, version, deliveryConfig, environment)
+      .also { result ->
+        if (!result) {
+          log.debug(
+            "{} constraint prevents promotion of {} {} to {}'s {} environment",
+            supportedType.name,
+            artifact.name,
+            version,
+            deliveryConfig.application,
+            environment.name
+          )
+        }
```

Comment:
It was driving me bananas that I couldn't see _which_ constraints were blocking promotion. At least with this we can find it in logs. We really need to figure out a way to enhance the API response for environments, but I figure that would make sense as a separate PR.